Tmedia 671 caption component

JackHowa opened the pull request:
Ticket: https://arcpublishing.atlassian.net/jira/software/c/projects/TMEDIA/boards/875?modal=detail&selectedIssue=TMEDIA-671&assignee=5e387d6a1b1d910e5dfd7a9b

# Ac: 

Create a <Caption> component

Uses props of Title, Caption and Credit that can be hidden at the block level

May have HTML in it which should render (i.e. `<b>, <i>`)

Have Scss file that adds the hooks to the Sass library

Includes three component properties: Title, Caption, and Credit

Caption - Title

Caption - Caption

Caption - Credit

Create a utility function to handle formatting credits that can be used by consumers - but not built into this component - [engine-theme-sdk/index.tsx at arc-themes-release-version-1.20 · WPMedia/engine-theme-sdk](https://github.com/WPMedia/engine-theme-sdk/blob/arc-themes-release-version-1.20/src/components/ImageMetadata/index.tsx#L47) 

Have test coverage

Have storybook documentation

# Test: 

1. Run `npm run storybook`
2. Go to Caption component 
3. See effect of html passed in via story `Renders html passed into title or caption` 
4. Look at how credit can take in multiple affiliations and/or bylines with stories


## Feedback 

- @nschubach's suggestion of using `.toHaveClass` was not doable as there was no way for me to select figcaption without role or without any text https://github.com/WPMedia/arc-themes-components/pull/29#pullrequestreview-896923800
- Removed out of scope sanitize if security is not a concern https://github.com/WPMedia/arc-themes-components/pull/29#discussion_r819742123
- Used dangerously set html based on feedback and requirements https://github.com/WPMedia/arc-themes-components/pull/29#discussion_r817198643
- Naming standards discussion from before

JackHowa committed:
feat: Add setup for caption component including taking in nodes as props for each field

JackHowa committed:
fix: Use fallback null rather than using &&

JackHowa committed:
feat: Dangerously set html what's passed in

JackHowa committed:
Merge branch 'arc-themes-release-version-1.20' into TMEDIA-671-caption-component

JackHowa committed:
docs: Based on feedback, this is an ok way of testing with react testing library. I couldn't target based on the text though based on the example

Couldn't target figcaption alone https://github.com/WPMedia/arc-themes-components/pull/29#discussion_r817213678

JackHowa committed:
docs: Allow taking in html for title and caption

JackHowa committed:
docs: Rename html props for clarity

JackHowa committed:
feat: Add credit format util

matthewroach commented on src/components/caption/index.jsx:
```diff
@@ -0,0 +1,35 @@
+const COMPONENT_CLASS_NAME = "c-caption";
+
+const Caption = ({ caption, className, credit, title }) => {
+	const formattedCredit = formatCredits(credit);
```
It was discussed that this utility would be for consumers to use and not directly built into the component

JackHowa replied:
updated!

nschubach commented on src/components/caption/index.test.jsx:
```diff
@@ -0,0 +1,84 @@
+
+		expect(screen.queryByText("title")).not.toBeNull();
+		// expect(screen.queryByText({{ by: [{ name: "bbbyyy" }], affiliation: [{ name: "affff" }] }})).not.toBeNull();
+		// expect(screen.queryByText("caption")).not.toBeNull();
```
Remove?

JackHowa replied:
updated

nschubach commented on src/components/caption/index.jsx:
```diff
@@ -0,0 +1,35 @@
+			className={className ? `${COMPONENT_CLASS_NAME} ${className}` : `${COMPONENT_CLASS_NAME}`}
+		>
+			{title ? (
+				<span dangerouslySetInnerHTML={{ __html: `${title} ` }} className="c-caption__title" />
```
Should be perhaps use the const for this?
```suggestion
				<span dangerouslySetInnerHTML={{ __html: `${title} ` }} className=`${COMPONENT_CLASS_NAME}__title` />
```

nschubach commented on src/components/caption/index.jsx:
```diff
@@ -0,0 +1,35 @@
+				<span dangerouslySetInnerHTML={{ __html: `${title} ` }} className="c-caption__title" />
+			) : null}
+			{caption ? (
+				<span dangerouslySetInnerHTML={{ __html: `${caption} ` }} className="c-caption__caption" />
```
```suggestion
				<span dangerouslySetInnerHTML={{ __html: `${caption} ` }} className=`${COMPONENT_CLASS_NAME}__caption` />
```

nschubach commented on src/components/caption/index.jsx:
```diff
@@ -0,0 +1,35 @@
+			{caption ? (
+				<span dangerouslySetInnerHTML={{ __html: `${caption} ` }} className="c-caption__caption" />
+			) : null}
+			{formattedCredit ? <span className="c-caption__credit">{formattedCredit}</span> : null}
```
```suggestion
			{formattedCredit ? <span className=`${COMPONENT_CLASS_NAME}__credit`>{formattedCredit}</span> : null}
```

nschubach commented on src/components/caption/index.stories.mdx:
```diff
@@ -0,0 +1,168 @@
+import { Meta, Story, Preview, Props } from "@storybook/addon-docs";
```
`Preview` and `Props` were renamed to `Canvas` and `ArgsTable` respectively.  Can you update this?

JackHowa replied:
updated! thanks @nschubach

brentmiller1973 commented:
@JackHowa This is very nitpicky, but in story examples where we demonstrate that HTML can be embedded, using divs or breaks are not immediately apparent.  `<b>` or `<strong>` tags give that feedback better to the user.  What do you think to modifying .storybook/preview-head.html to inject some styles?  So for `<b` just update the file to:

```
<!-- .storybook/preview-head.html -->
<style>
    b {
        font-weight: bold;
    }
</style>
```

Then you will get a better demonstration in the "Renders html passed into title or caption" story:

![image](https://user-images.githubusercontent.com/833461/156841876-5b9356a8-6d47-4f10-9e5f-d964f4b67716.png)

Or as an alternative, just inline the style in the story:

```
<Preview>
	<Story name="Renders html passed into title or caption">
		<Caption
			title='<p><b style="font-weight:bold">Important Alert</b></p>'
			caption="There is man on the loose biting dogs</div><div>All dogs should evacuate the area"
			credit={{ by: [{ name: "John Doe" }], affiliation: [{ name: "Stock Photo" }] }}
		/>
	</Story>
</Preview>
```

nschubach reviewed (dismissed):
Class requests, Storybook Doctag name updates, and some commented code to remove.

matthewroach commented:
@visionbegin - To your comment about setting default values for bold. We should do this inside of the Sass library, maybe we are resetting too much. Think like `<b>` and `<em>` should have the bold and italic CSS set out of the box - why should we remove it? or even make additional work for ourselves to have to redefine it?

JackHowa committed:
refactor: Update the format functionality based off of util feedback

https://github.com/WPMedia/arc-themes-components/pull/35#pullrequestreview-900689232

Commit:
Update src/components/caption/index.jsx

Co-authored-by: nschubach <nschubach@gmail.com>

Commit:
Update src/components/caption/index.jsx

Co-authored-by: nschubach <nschubach@gmail.com>

JackHowa committed:
style: Fix formatting based off of Nick's suggestion of constant reusage

https://github.com/WPMedia/arc-themes-components/pull/35#discussion_r819915229

JackHowa committed:
Merge branch 'arc-themes-release-version-1.20' into TMEDIA-671-caption-component

brentmiller1973 commented:
> @visionbegin - To your comment about setting default values for bold. We should do this inside of the Sass library, maybe we are resetting too much. Think like `<b>` and `\<em\>` should have the bold and italic CSS set out of the box - why should we remove it? or even make additional work for ourselves to have to redefine it?

@matthewroach I'm not thinking so much about our Sass settings as I am thinking about providing immediate feedback to the client visually in the story.  While my initial suggestion was to modify the storybook style header, I still think there is value to inline the styles on some stories (especially html injection) to demonstrate this.

JackHowa commented:
is this new? `npm run storybook`

```
37% building 230/243 modules 13 active ...s/node_modules/core-js/modules/esnext.symbol.metadata.jsWarning: components "paragraph" is configured with no options
    src/scss/root/_index.scss 68:4   token-map-to-css-variable()
    src/scss/root/_index.scss 98:4   root-tokens()
    src/scss/root/_index.scss 132:4  @forward
    src/scss/index.scss 1:1          @forward
    scss.scss 1:1                    @use
    .storybook/theme.scss 1:1        root stylesheet

```

JackHowa committed:
refactor: Rename storybook imports

Thanks https://github.com/WPMedia/arc-themes-components/pull/35#discussion_r819915812

JackHowa commented:
updated based on feedback!

JackHowa commented:
Just for reference, the css reset reset the font-style. That's why it doesn't look as expected: 

```
	&,
	&::before,
	&::after {
		font: inherit;
```

cc @visionbegin 
![Screen Shot 2022-03-04 at 15 46 29](https://user-images.githubusercontent.com/5950956/156846184-feeb20a3-6608-4971-82ed-e13f09de3d4e.png)

JackHowa commented:
I guess this gets back to the discussion of adding "example" or something themes to the storybook picker. I'm not sure if this conversation is relevant to this pr

matthewroach reviewed (dismissed):
The separate util function looks good, but there's no documentation for it that's accessible to developers. We can add stories for these utils too, and use the grouping feature in storybook to put them under a Utilities grouping

matthewroach commented:
> is this new? `npm run storybook`
> 
> ```
> 37% building 230/243 modules 13 active ...s/node_modules/core-js/modules/esnext.symbol.metadata.jsWarning: components "paragraph" is configured with no options
>     src/scss/root/_index.scss 68:4   token-map-to-css-variable()
>     src/scss/root/_index.scss 98:4   root-tokens()
>     src/scss/root/_index.scss 132:4  @forward
>     src/scss/index.scss 1:1          @forward
>     scss.scss 1:1                    @use
>     .storybook/theme.scss 1:1        root stylesheet
> ```

This is due to the `.storybook/theme.scss` file having a reference for `paragraph: ()` and nothing set - the warnings is built into our Sass library to provide developers additional feedback

JackHowa committed:
docs: Remove unused style that causes warnings

JackHowa committed:
docs: Add utilities for format credits documentation along with properties and usage

JackHowa commented:
> The separate util function looks good, but there's no documentation for it that's accessible to developers. We can add stories for these utils too, and use the grouping feature in storybook to put them under a Utilities grouping

updated @matthewroach

JackHowa commented:
> > is this new? `npm run storybook`
> > ```
> > 37% building 230/243 modules 13 active ...s/node_modules/core-js/modules/esnext.symbol.metadata.jsWarning: components "paragraph" is configured with no options
> >     src/scss/root/_index.scss 68:4   token-map-to-css-variable()
> >     src/scss/root/_index.scss 98:4   root-tokens()
> >     src/scss/root/_index.scss 132:4  @forward
> >     src/scss/index.scss 1:1          @forward
> >     scss.scss 1:1                    @use
> >     .storybook/theme.scss 1:1        root stylesheet
> > ```
> 
> This is due to the `.storybook/theme.scss` file having a reference for `paragraph: ()` and nothing set - the warnings is built into our Sass library to provide developers additional feedback

updated this as well to remove the warning

nschubach reviewed (approved):
Looks like all the concerns were addressed and I do not see any issue with the storybook or test coverage.
